### ISSUE TITLE
Local changes on nested model not acknowledged.

### ISSUE THREAD

**Issue by ygongdev** [hjdivad/ember-m3#217]
If I understand correctly, according to the documentation for `ember-data`,
`_data holds all of the attributes that have been acknowledged by a backend via the adapter`

In this scenario where we change the nested model. After the save promise resolves, e.g `save().then(resolvedModel => ...)`,
`resolvedModel._internalModel._recordData._data` should contain the changes made on its nested model. But that doesn't seem to be the case because I kept seeing the old _data.
The `childRecordData`, e.g `resolvedModel._internalModel._recordData._childRecordData.[key][index]` contains the changes in its `_attributes`.

To summarize, it seems like that currently if we make a change to a nested model, the backend is not acknowledging the changes properly.

**Comment by ygongdev** [comment]
Currently, it seems like the adapter/backend/server? only cares about the top-level model and doesn't do anything about the nested models. 

1. Also in this scenario, we don't have recursive `willCommit` on the child records because we fail the `isArray` check. 
2. Even if we pass the check, recursive `willCommit` doesn't inform its parent of anything, it only sits `attributes` and stuff on itself.
3. Because the adapter/backend/server? doesn't "know" about the nested changes, when we do `didCommit` on the top-level model, there's nothing from the server, so `_mergeUpdates` is never called.

Perhaps instead of relying on the recursive approach, we can perform a breadth-first on all the nested models and assign their changes to the top level model, so when `adapterWillCommit` is called, it has all the changes it needs to the server.

Then inside `didCommit`, we can do another breath-first on all the nested models to "clean" them up, so there aren't any leftover `_inFlightAttributes` and etc.

**Comment by hjdivad** [comment]
Thanks for digging into this @ygongdev.

The issue is that nested models in arrays are treated differently than nested models outside of arrays.  The reason for this is there isn't in general a way to merge updates from the server because you can't cross-reference new array items to existing ones in general.

What should happen when a record has some key `K` that is an array of nested models that have changed attributes is

1. If the server includes `K` in the response, destroy the array of child record data without committing their local changes and use whatever data the server provided
2. If the server does *not* include `K` in the response, `_mergeUpdates` should call `didCommit` on the items within the array and merge their `inFlightAttributes` back into its `data[K]` at the right index

And finally, for the case where the attribute is *not* an array
1. If the server does not include the attribute, call `didCommit` on the nested record data and merge its `inFlightAttributes` into `data[K]`

**Comment by ygongdev** [comment]
Fix was merged. Will re-open if the fix doesn't solve the issue.
